Resource condition returns true if the request's resouce contains the context string

devdavidkarlsson opened the pull request:
We need to validate that the context we are currently in is reflected in the resource string.

I.e. a user in contextX can only get resources in contextX
Hence we need a context->resource evaluation.

Please have mercy on our PR :)

arekkas commented on condition_resource.go:
```diff
@@ -0,0 +1,28 @@
+		return false
+	}
+
+	filter += ":"
```
This delimiter should be configurable

devdavidkarlsson replied:
ok

arekkas commented on condition_resource.go:
```diff
@@ -0,0 +1,28 @@
+)
+
+// ResourceCondition is a condition which is fulfilled if the request's subject is equal to the given value string
+type ResourceCondition struct{}
```
This goes more in the direction of "the context must match a substring within the resource name" right?

devdavidkarlsson replied:
yes!

arekkas replied:
Do you have an idea how the name of the struct could represent this?

devdavidkarlsson replied:
ResourceContainsCondition?
ResourceStringContainsCondition?

devdavidkarlsson commented:
Readme and delimiter added.

arekkas commented on condition_resource_test.go:
```diff
@@ -0,0 +1,48 @@
+		{delimiter: ":", value: "GroupId:2", resource: "abc", pass: false},
+		{delimiter: ":", value: "", resource: "abc", pass: false},
+		{delimiter: "", value: "test", resource: "abc", pass: false},
+	} {
```
The tests are not really expressive at the moment, I'm not 100% sure what should fail and what shouldn't. For example, should this fail?

* {delimiter: ":", value: "GroupId:2", resource: "resources:sensor-data:GroupId:223:deviceIds:a9b576e8-7419-4eed-a010-7f68ec0ff588"},
* {delimiter: ":", value: "GroupId:2", resource: "resources:sensor-data:GroupId:2"},

If so, please add more tests to express this :)

devdavidkarlsson replied:
I am afraid (but not certain) that the regex could be a lot slower in comparison to the contains, since we need to do this resource check on a lot of requests. Did you take that into consideration for condition string match?

https://awmanoj.github.io/tech/2016/09/08/on-slow-regular-expressions-golang/

devdavidkarlsson replied:
210µs vs 40µs on average for the test cases... i.e. more than 5 times slower...

arekkas replied:
Regex is slower (it has O(n) in Go) than simple string comparison.

What I fear about this PR is that it is aimed quite specific at one use case (could you please take the time to answer the question regarding what fails and wont so I get a better understanding of this) which could be generalized with regular expressions.

If you have fewer than 10000 requests per second (~1B requests per day) I wouldn't worry about this too much ;)

devdavidkarlsson replied:
Yes sure will, thanks for your answer, good points. regex should be used for making this functionality more generic and usable...

arekkas commented on README.md:
````diff
@@ -475,6 +476,51 @@ var err = warden.IsAllowed(&ladon.Request{
+The Condition requires a value string and delimiter (needs to match the resource string) to be passed.
+
+```go
+var err = warden.IsAllowed(&ladon.Request{
````
Does this pass? If so please add that info

arekkas commented on README.md:
```diff
@@ -475,6 +476,51 @@ var err = warden.IsAllowed(&ladon.Request{
+
+Checks if the string value passed in the access request's context is present in the resource string.
+
+The Condition requires a value string and delimiter (needs to match the resource string) to be passed.
```
Could you explain more how the delimiter is used here?

devdavidkarlsson commented:
Updated the readme, switched to the regex eval., added trivial regex cases to show the ".*" functionality  and added some comments around the tests to make them easier to follow...

:)

devdavidkarlsson commented:
Testing it out in a warden instance atm, some more changes to come..

arekkas commented on README.md:
```diff
@@ -46,6 +46,7 @@ ORY builds solutions for better internet security and accessibility. We have a c
       - [String Match Condition](#string-match-condition)
       - [Subject Condition](#subject-condition)
       - [String Pairs Equal Condition](#string-pairs-equal-condition)
+      - [Resource Condition](#resource-condition)
```
This must be updated :)

arekkas commented on README.md:
```diff
@@ -475,6 +476,58 @@ var err = warden.IsAllowed(&ladon.Request{
+
+The Condition requires a value string and delimiter (needs to match the resource string) to be passed.
+
+A resource could for instance be: `myrn:some.domain.com:resource:123` and `myrn:some.otherdomain.com:resource:123` (the `:` is then considered a delimiter, and used by the condition to be able to separate the resource components from each other) to allow an action to the resources on `myrn:some.otherdomain.com` you could for instance create a resource condition  with
```
I would explain here that the part of the resource "path" must exactly match what you're searching for. So `foo:bar` matches `foo:bar` but not `foo:bara` nor `foo:bara:baz`

devdavidkarlsson replied:
Todo resource "matches" you only need to input the "full string" in contains, so it makes no use splitting in two different cases (assuming the cost of regex is neglible as you mentioned)...

Not sure how it would work with a resource definition <[^:]+>:foo:bar:<[^:]+>. I always saw resource as an object (singular). 

Can you use regex in the middle of the resource definition like that currently?

`rn:myorg:group:*:device:*`

How would you get device:22 and not device:222 using regex?




Here is an example of the resource condition in Warden:
```
curl -X POST \
  http://localhost:8080/policies \
  -H 'Cache-Control: no-cache' \
  -H 'Content-Type: application/json' \
  -H 'Postman-Token: d279f81a-f1a0-d341-cfea-2816607c00a0' \
  -d '        {
          "description": "One policy to rule them all.",
          "subjects": ["users:peter"],
          "actions" : ["delete"],
          "effect": "allow",
          "resources": [
            "resources:tenant:2"
          ],
          "conditions": {
            "resourceFilter": {
                "type": "ResourceContainsCondition"
            }
          }
        }'
```

```
curl -X POST \
  http://localhost:8080/warden/allowed \
  -H 'Cache-Control: no-cache' \
  -H 'Content-Type: application/javascript' \
  -H 'Postman-Token: 590e4962-c1e0-93dc-3aaf-2133607d6cc9' \
  -d '        {
          "subject": "users:peter",
          "action" : "delete",
          "resource": "resources:tenant:2",
          "context": {
            "resourceFilter": {
            	"Value": "tenant:2"
            	
            }
          }
        }'
```

arekkas replied:
> Can you use regex in the middle of the resource definition like that currently?
>
> rn:myorg:group:*:device:*

`rn:myorg:group:<.*>:device:<.*>`

> How would you get device:22 and not device:222 using regex?

I can't give you a full tutorial on regex, but doing stuff like `device:22[^\:]+` is the way to go.

In your example, I don't really get the need for checking for `tenant:2` as the policy itself is already doing that (it must match `"resources:tenant:2"`). Why do you need a second thing (condition) to solve what the resource part of the policy definition itself is already capable of? Maybe you could explain your use case?

devdavidkarlsson replied:
please refer to my first comment on this pr.

arekkas replied:
Ah sorry, the discussion has been going for a while that I forgot about it :) So instead of having tons of policies reflecting that, you simply want to do that using the context - that makes sense IMO.

The main issue with the current condition is that you're assuming that we have delimiters while in fact this might not be the case for some people. So instead it would be more useful to match this in total with a regular expression (leave the delimiter out completely). The example above would be covered with `(^|:.*)tenant:2($|:.*)` (this matches `tenant:2`, `tenant:2:123`, `123:tenant:2:123` but not `tenant:22`). Obviously, as you know the layout of your resource names you probably won't even need that and simply checking for `:tenant:2:` or `resources:tenant:2` would already be enough (no regexp required).

I suggest to throw out the delimiter part and either have one or two conditions instead:

* ResourceContains (uses strings.Contains())
* ResourceMatches (uses regexp.MatchString())

arekkas commented on condition_resource_contains_test.go:
```diff
@@ -0,0 +1,58 @@
+		pass      bool
+	}{
+		//Correct matching within resource string:
+		{matches: "Exact resource component match in resource string, with : delimiter", delimiter: ":", value: "GroupId:2", resource: "resources:sensor-data:GroupId:2:deviceIds:a9b576e8-7419-4eed-a010-7f68ec0ff588", pass: true},
```
```
	r := new(ResourceContainsCondition)
	fmt.Printf("got: %b", r.Fulfills(map[string]interface{}{"Value": "foo:bar", "Delimiter": ":"}, &Request{
		Resource: "ffoo:bar",
	}))
```

returns true which is imho a false positive

devdavidkarlsson replied:
Yes, good find, postfixing with a delimiter is not going to be enough, prefixing in a commit pushed now.

arekkas commented on condition_resource_contains.go:
```diff
@@ -0,0 +1,40 @@
+		return false
+	}
+
+	valueString, ok := filter["Value"].(string)
```
For consistency, these must be lower case (`value` and `delimiter`)

devdavidkarlsson commented on condition_resource_contains_test.go:
```diff
@@ -26,6 +26,15 @@ func TestResourceContains(t *testing.T) {
+		{matches: "Exact resource component match in resource string, with no delimiter", value: "(notMyResource$)", resource: "myResource", pass: false},
+		{matches: "Exact resource component match with resource string, with no delimiter", value: "(^myResource$)", resource: "notMyResource", pass: false},
+		{matches: "Exact resource component match with resource string, with no delimiter", value: "(^myResource$)", resource: "MyResourceNot", pass: false},
+		{matches: "Exact resource component match with resource string, with no delimiter", value: "(^myResource$)", resource: "myResource", pass: true},
```
@arekkas I made the delimiter optional, so one could choose to use that or go full-blown regex as shown here.
This could make sense if resource strings are not composed of delimited values...

arekkas replied:
That makes sense! I'm giving it one last pass and then, I think, we're good to go :)

arekkas commented on condition_resource_contains.go:
```diff
@@ -0,0 +1,41 @@
+
+	// Append delimiter to strings to prevent delim+1 being interpreted as delim+10 being present
+	filterValue := delimiterString + valueString + delimiterString
+	resourceString := r.Resource + delimiterString
```
This will not work if the resource is `foo:bar`, delimiter is `:`, and the filter value is also `foo:bar`. Add this as a failing test case before working on it.

devdavidkarlsson replied:
Will have a look on it tomorrow, 
Note to self: an exakt string equals check before the contains check, may solve this.

arekkas commented on condition_resource_contains.go:
```diff
@@ -0,0 +1,41 @@
+	filterValue := delimiterString + valueString + delimiterString
+	resourceString := r.Resource + delimiterString
+
+	matches, _ := regexp.MatchString(filterValue, resourceString)
```
Replace with `strings.Contains`, it makes more sense here

arekkas commented on condition_resource_contains_test.go:
```diff
@@ -0,0 +1,71 @@
+		{matches: "Exact resource component match in resource string, with no delimiter", value: ":GroupId:2:", resource: "resources:sensor-data:GroupId:2:deviceIds:a9b576e8-7419-4eed-a010-7f68ec0ff588", pass: true},
+		{matches: "Exact resource component match in resource string, with no delimiter", value: "myResource", resource: "myResource", pass: true},
+		//Better to use regex for string, to avoid false positives:
+		{matches: "Exact resource component match in resource string, with no delimiter", value: "(notMyResource$)", resource: "myResource", pass: false},
```
Remove regex tests, they overcomplicate this condition - we should add another one if required that is capable of doing regex

arekkas commented on condition_resource_contains_test.go:
```diff
@@ -0,0 +1,71 @@
+		pass      bool
+	}{
+		//Correct matching within resource string without delimiter:
+		{matches: "Exact resource component match in resource string, with no delimiter", value: ":GroupId:2:", resource: "resources:sensor-data:GroupId:2:deviceIds:a9b576e8-7419-4eed-a010-7f68ec0ff588", pass: true},
```
There are test cases missing here. I would like to have the following covered:

* With delimiter (`:`):
  * Tests if it passes when the value is a prefix (value: `foo:bar`, resource: `foo:bar:baz`)
  * Tests if it fails when the value is not a prefix  (value: `foo:bar`, resource: `foo:baz:bar`)
  * Tests if it passes when the value is an affix (value: `foo:bar`, resource: `baz:foo:bar`)
  * Tests if it fails when the value is not an affix (value: `foo:baz`, resource: `baz:foo:baz`)
  * Tests if it passes when the value is somewhere in the middle (value: `foo`, resource: `baz:foo:bar`)
  * Tests if it fails when the value is not somewhere in the middle (value: `foo`, resource: `baz:foo:baz`)
* Without delimiter -> same test cases as above but without delimiter, obviously :)

We should also decide what happens when `{delimiter: ":", value: ":foo:", resource: "bar:foo:bar"`. Right now this would fail. Should it?

Everything else can be thrown out - unless I missed something.

devdavidkarlsson replied:
much obliged, will update.

devdavidkarlsson replied:
Tests if it fails when the value is not somewhere in the middle (value: foo, resource: baz:foo:baz) is meant to be foo:bar I am guessing

arekkas replied:
Yes, the resource should be something like `baz:faa:baz` :)

arekkas commented on condition_resource_contains_test.go:
```diff
@@ -0,0 +1,71 @@
+		{matches: "Not exact resource component match at end of resource string, with _ delimiter", delimiter: "_", value: "deviceIds_a9b", resource: "resources_sensordata_GroupId_2_deviceIds_a9b6", pass: false},
+		{matches: "Not found in resource string", delimiter: ":", value: "GroupId:2", resource: "abc", pass: false},
+		//More advanced regex:
+		{matches: "Regex resource component match at end of resource string, with _ delimiter", delimiter: "_", value: "deviceIds_a9.*", resource: "resources_sensordata_GroupId_2_deviceIds_a9b", pass: true},
```
Remove regex test cases

devdavidkarlsson commented:
Should be alright now then. updated as specified, added a test I thought missing, showing the usage of the delimiter, plus test for missing value.

arekkas commented:
Thank you for being so patient and for your contribution!

arekkas commented:
One more thing, please sign the [DOC](https://github.com/ory/ladon/blob/master/CONTRIBUTING.md#developers-certificate-of-origin) by using `git commit --amend -s`, then I'll merge :)

devdavidkarlsson commented:
Signed :)

arekkas commented:
Thank you for your continued effort!
